### PR TITLE
[libc++] Fix uninitialized algorithms when using unconstrained comparison operators

### LINKED ISSUE #69334: [libc++] compile fails when using `std::unitialized_copy()` with an output iterator type that has `operator!=()` overload
## Description

When using `std::uninitialized_copy()` from libc++:

```cpp
template< class InputIt, class NoThrowForwardIt >
NoThrowForwardIt uninitialized_copy( InputIt first, InputIt last,
                                     NoThrowForwardIt d_first );
```

If the user overloaded `operator!=()` like this:

```cpp
template< class Iter >
bool operator!=( NoThrowForwardIt const& lhs, Iter const& rhs );
```

Compilation fails due to ambiguous overload resolution.

## Minimal reproducible example

<details>
<summary>main.cpp</summary>

```cpp
#include <array>
#include <iostream>
#include <iterator>
#include <memory>

namespace
{
template<typename T>
class myCrappyIterator
{
public:
    using value_type = typename std::iterator_traits<T*>::value_type;
    using difference_type = typename std::iterator_traits<T*>::difference_type;
    using reference = typename std::iterator_traits<T*>::reference;
    using pointer = typename std::iterator_traits<T*>::pointer;
    using iterator_category = typename std::iterator_traits<T*>::iterator_category;

    myCrappyIterator() = default;

    explicit myCrappyIterator(T* const ptr)
        : ptr_(ptr)
    {
    }

    myCrappyIterator(myCrappyIterator const& that) = default;
    myCrappyIterator(myCrappyIterator&& that) = default;
    myCrappyIterator<T>& operator=(myCrappyIterator const& that) = default;
    myCrappyIterator<T>& operator=(myCrappyIterator&& that) = default;

    constexpr reference operator*() const noexcept
    {
        return *ptr_;
    }

    constexpr pointer operator->() const noexcept
    {
        return std::addressof(*ptr_);
    }

    constexpr myCrappyIterator& operator++() noexcept
    {
        ++ptr_;
        return *this;
    }

    constexpr myCrappyIterator operator++(int) const noexcept
    {
        myCrappyIterator old{ *this };
        ++(*this);
        return old;
    }

    constexpr bool operator==(myCrappyIterator const& that) const noexcept
    {
        return this->ptr_ == that.ptr_;
    }

    constexpr bool operator!=(myCrappyIterator const& that) const noexcept
    {
        return !this->operator==(that);
    }

private:
    pointer ptr_;

    template<typename Iter>
    friend bool operator!=(myCrappyIterator const& lhs, Iter const& rhs) noexcept;
};

template<typename T, typename Iter>
constexpr bool operator!=(myCrappyIterator<T> const& lhs, Iter const& rhs) noexcept
{
    return lhs.ptr_ == rhs;
}
} // namespace

int main()
{
    static auto constexpr src = std::array{ 1, 2, 3, 4 };
    auto dst = decltype(src){};

    std::uninitialized_copy(std::begin(src), std::end(src), myCrappyIterator{ std::data(dst) });

    return 0;
}
```

</details>

clang console output:

```console
In file included from main.cpp:1:
In file included from /usr/include/c++/v1/array:114:
In file included from /usr/include/c++/v1/algorithm:667:
In file included from /usr/include/c++/v1/functional:506:
In file included from /usr/include/c++/v1/__functional/function.h:24:
In file included from /usr/include/c++/v1/memory:819:
In file included from /usr/include/c++/v1/__memory/ranges_uninitialized_algorithms.h:22:
/usr/include/c++/v1/__memory/uninitialized_algorithms.h:45:41: error: use of overloaded operator '!=' is ambiguous (with operand types '(anonymous namespace)::myCrappyIterator<int>' and 'std::__unreachable_sentinel')
    for (; __ifirst != __ilast && __idx != __olast; ++__ifirst, (void)++__idx)
                                  ~~~~~ ^  ~~~~~~~
/usr/include/c++/v1/__memory/uninitialized_algorithms.h:61:26: note: in instantiation of function template specialization 'std::__uninitialized_copy<int, const int *, const int *, (anonymous namespace)::myCrappyIterator<int>, std::__unreachable_sentinel>' requested here
  auto __result = _VSTD::__uninitialized_copy<_ValueType>(_VSTD::move(__ifirst), _VSTD::move(__ilast),
                         ^
main.cpp:82:10: note: in instantiation of function template specialization 'std::uninitialized_copy<const int *, (anonymous namespace)::myCrappyIterator<int>>' requested here
    std::uninitialized_copy(std::begin(src), std::end(src), myCrappyIterator{ std::data(dst) });
         ^
/usr/include/c++/v1/__memory/uninitialized_algorithms.h:30:55: note: candidate function [with _Iter = (anonymous namespace)::myCrappyIterator<int>]
  _LIBCPP_HIDE_FROM_ABI friend _LIBCPP_CONSTEXPR bool operator!=(const _Iter&, __unreachable_sentinel) _NOEXCEPT {
                                                      ^
main.cpp:67:17: note: candidate function [with Iter = std::__unreachable_sentinel]
    friend bool operator!=(myCrappyIterator const& lhs, Iter const& rhs) noexcept;
                ^
main.cpp:71:16: note: candidate function [with T = int, Iter = std::__unreachable_sentinel]
constexpr bool operator!=(myCrappyIterator<T> const& lhs, Iter const& rhs) noexcept
               ^
1 error generated.
```

### PR DESCRIPTION
**Pull request by ldionne**
If an iterator passed to std::uninitialized_copy & friends provided an unconstrained comparison operator, we would trigger an ambiguous overload resolution because we used to compare against __unreachable_sentinel in our implementation.

This patch fixes that by only comparing the output iterator when it is actually required, i.e. in the <ranges> versions of the algorithms.

Fixes #69334

### CODE REVIEW

**Review by phyBrackets — COMMENTED** [review]
Thanks for the quick patch!

File: libcxx/include/__memory/uninitialized_algorithms.h [hunk 1]
```diff
@@ -44,26 +44,23 @@
...
-                     _ForwardIterator __ofirst, _Sentinel2 __olast) {
+template <class _ValueType, class _InputIterator, class _Sentinel1, class _ForwardIterator, class _OutputIterCompare>
+inline _LIBCPP_HIDE_FROM_ABI pair<_InputIterator, _ForwardIterator> __uninitialized_copy(
+    _InputIterator __ifirst, _Sentinel1 __ilast, _ForwardIterator __ofirst, _OutputIterCompare __out_iter_at_end) {
   _ForwardIterator __idx = __ofirst;
 #ifndef _LIBCPP_HAS_NO_EXCEPTIONS
   try {
 #endif
-    for (; __ifirst != __ilast && __idx != __olast; ++__ifirst, (void)++__idx)
+    for (; __ifirst != __ilast && !__out_iter_at_end(__idx); ++__ifirst, (void)++__idx)
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__memory/uninitialized_algorithms.h:63]: Previously we would compare against a sentinel that always said `false` when asked "is the output iterator at end". After the patch we just call a predicate that returns `false` directly, so those are equivalent (but we don't use `operator!=` or `operator==` like we used to).

File: libcxx/include/__memory/ranges_uninitialized_algorithms.h [hunk 2]
```diff
@@ -232,8 +233,8 @@ struct __fn {
   operator()(_InputIterator __ifirst, iter_difference_t<_InputIterator> __n,
              _OutputIterator __ofirst, _Sentinel __olast) const {
     using _ValueType = remove_reference_t<iter_reference_t<_OutputIterator>>;
-    auto __result = _VSTD::__uninitialized_copy_n<_ValueType>(_VSTD::move(__ifirst), __n,
-                                                              _VSTD::move(__ofirst), _VSTD::move(__olast));
+    auto __stop      = [&__olast](auto&& __out_iter) { return __out_iter == __olast; };
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/include/__memory/ranges_uninitialized_algorithms.h:236]: Yeah, the current names are really not great.

**Review by phyBrackets — APPROVED** [review]
Thanks for fixing, It looks fine now!

**Pull request merged**